src/TinyUPnP.cpp

```diff
@@ -229,61 +232,58 @@ UpdateState TinyUPnP::updatePortMapping(unsigned long intervalMs, callback_funct
 	return NOP;
 }
 
-boolean TinyUPnP::testConnectivity(long startTime) {
-	IPAddress gatewayIP = WiFi.gatewayIP();
-	debugPrint(F("Testing WiFi connection for ["));
-	debugPrint(ipAddressToString(WiFi.localIP()));
-	debugPrint("]");
+boolean TinyUPnP::testConnectivity(unsigned long startTime) {
+	//IPAddress gatewayIP = WiFi.gatewayIP();
+	UPNP_DEBUG(F("Testing WiFi connection for ["));
+	UPNP_DEBUG(ipAddressToString(WiFi.localIP()));
+	UPNP_DEBUG("]");
 	while (WiFi.status() != WL_CONNECTED) {
 		if (_timeoutMs > 0 && startTime > 0 && (millis() - startTime > _timeoutMs)) {
-			debugPrint(F(" ==> Timeout expired while verifying WiFi connection"));
+			UPNP_DEBUG(F(" ==> Timeout expired while verifying WiFi connection"));
 			_wifiClient.stop();
 			return false;
 		}
 		delay(200);
-		debugPrint(".");
+		UPNP_DEBUG(".");
 	}
-	debugPrintln(" ==> GOOD");  // \n
+	UPNP_DEBUGln(" ==> GOOD");  // \n
 
-	debugPrint(F("Testing internet connection"));
+	UPNP_DEBUG(F("Testing internet connection"));
 	_wifiClient.connect(connectivityTestIp, 80);
 	while (!_wifiClient.connected()) {
 		if (startTime + TCP_CONNECTION_TIMEOUT_MS > millis()) {
-			debugPrintln(F(" ==> BAD"));
+			UPNP_DEBUGln(F(" ==> BAD"));
 			_wifiClient.stop();
 			return false;
 		}
 	}
 
-	debugPrintln(F(" ==> GOOD"));
+	UPNP_DEBUGln(F(" ==> GOOD"));
 	_wifiClient.stop();
 	return true;
 }
 
 boolean TinyUPnP::verifyPortMapping(gatewayInfo *deviceInfo) {
-	debugPrintln(F("Verifying rule in IGD"));
+	UPNP_DEBUGln(F("Verifying rule in IGD"));
 
 	// connect to IGD (TCP connection) again, if needed, in case we got disconnected after the previous query
 	unsigned long timeout = millis() + TCP_CONNECTION_TIMEOUT_MS;
 	if (!_wifiClient.connected()) {
 		while (!connectToIGD(_gwInfo.host, _gwInfo.actionPort)) {
 			if (millis() > timeout) {
-				debugPrintln(F("Timeout expired while trying to connect to the IGD"));
+				UPNP_DEBUGln(F("Timeout expired while trying to connect to the IGD"));
 				_wifiClient.stop();
 				return false;
 			}
 			delay(500);
 		}
 	}
+	String request = F("<?xml version=\"1.0\"?>\r\n<s:Envelope xmlns:s=\"http://schemas.xmlsoap.org/soap/envelope/\" s:encodingStyle=\"http://schemas.xmlsoap.org/soap/encoding/\">\r\n<s:Body>\r\n<u:GetSpecificPortMappingEntry xmlns:u=\"urn:schemas-upnp-org:service:WANPPPConnection:1\">\r\n<NewRemoteHost></NewRemoteHost>\r\n<NewExternalPort>");
+	request += String(_rulePort);
+	request += F("</NewExternalPort>\r\n<NewProtocol>");
+	request += _ruleProtocol;
+	request += F("</NewProtocol>\r\n</u:GetSpecificPortMappingEntry>\r\n</s:Body>\r\n</s:Envelope>\r\n");
```

Comment:
let's leave this in second phase please.
I am not fond of using String, and I worked hard to avoid this in order to make the package more stable.